Src/IronPython/Runtime/List.cs

```diff
@@ -91,23 +91,21 @@ public class List : IList, ICodeFormattable, IList<object>, IReversible, IStruct
         }
 
         public void __init__(CodeContext context, object sequence) {
+            int len = INITIAL_SIZE; 
             try {
-                object len;
-                if (PythonTypeOps.TryInvokeUnaryOperator(context, sequence, "__len__", out len)) {
-                    int ilen = context.LanguageContext.ConvertToInt32(len);
-                    _data = new object[ilen];
-                    _size = 0;
-                    extend(sequence);
-                } else {
-                    _data = new object[INITIAL_SIZE];
-                    _size = 0;
-                    extend(sequence);
+                object len_obj;
+                if (PythonTypeOps.TryInvokeUnaryOperator(context, sequence, "__len__", out len_obj) ||
+                    PythonTypeOps.TryInvokeUnaryOperator(context, sequence, "__length_hint__", out len_obj)) {
+                    if (!(len_obj is NotImplementedType)) {
```

Comment:
You have this `NotImplementedType` check in some places and not in others. Is this something that should be checked all the time? Maybe we should generalize this into a `PythonOps` method? Maybe something like:
```C#
public bool TryInvokeLengthHint(CodeContext context, object sequence, out int hint);
```